src/rust/engine/src/nodes.rs

```diff
@@ -868,13 +906,17 @@ impl WrappedNode for Task {
     let func = self.task.func;
     let entry = self.entry;
     let product = self.product;
+    let can_modify_workunit = self.task.can_modify_workunit;
 
     let result_val = externs::call(&externs::val_for(&func.0), &deps)?;
     match externs::get_type_for(&result_val) {
-      t if t == context.core.types.coroutine => {
-        Self::generate(context, params, entry, result_val).await
+      t if t == context.core.types.coroutine => Self::generate(context, params, entry, result_val)
+        .await
+        .map(|v| (v, None)),
+      t if t == product => {
+        let maybe_new_level = Self::compute_new_workunit_level(can_modify_workunit, &result_val);
```

Comment:
Important: This won't run if we end up using a generator: will need to move the `compute_new_workunit_level` to somewhere after this block so that it runs on the result in both cases.